### PR TITLE
A0-2982: Synchronize missing imports

### PR DESCRIPTION
**Pull request by timorl**
# Description

Handler now remembers that it should attempt reacquiring information about imported blocks that were too new.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- I have added tests
- I have created new documentation

### CODE REVIEW

File: finality-aleph/src/sync/mock/backend.rs [hunk 1]
```diff
@@ -114,15 +122,24 @@ impl Backend {
...
+                || is_predecessor(
+                    &storage.blockchain,
+                    id,
+                    top_finalized_id,
+                    &to_prune,
+                    &definitely_correct,
+                )
+            {
+                definitely_correct.insert(id.clone());
+                continue;
```
Comment (ggawryal) [inline finality-aleph/src/sync/mock/backend.rs:138]: Nit: remove continue?

File: finality-aleph/src/sync/handler/mod.rs [hunk 2]
```diff
@@ -82,6 +83,90 @@ pub trait HandlerTypes {
...
+            last_sync,
+        } = self
+        {
+            let top_finalized = chain_status.top_finalized()?.header().id();
+            // we don't want this to happen too often, but it also cannot be too close to the max forest size, thus semi-random weird looking threshold
+            if *last_sync + 1312 >= top_finalized.number() {
+                return Ok(());
+            }
+            let mut to_import = VecDeque::from(chain_status.children(top_finalized.clone())?);
+            while let Some(header) = to_import.pop_front() {
```
Comment (ggawryal) [inline finality-aleph/src/sync/handler/mod.rs:151]: This loop will execute pessimistically 1312 + a few times in presence of forks, right? Are we sure this is not too many synchronous operations?

File: finality-aleph/src/sync/handler/mod.rs [hunk 3]
```diff
@@ -82,6 +83,90 @@ pub trait HandlerTypes {
...
+                if header.id().number() > *highest_missed {
+                    break;
+                }
+                // we suppress all errors except `TooNew` since we are likely trying to mark things that are already marked and they would be throwing a lot of stuff
```
Comment (ggawryal) [inline finality-aleph/src/sync/handler/mod.rs:155]: Maybe at least `ParentNotImported` and `HeaderMissingParentId` could not be suppressed?

**Review by ggawryal — COMMENTED** [review]
Looks good, left a few comments for clarity.

File: finality-aleph/src/sync/handler/mod.rs [hunk 4]
```diff
@@ -1790,4 +1890,35 @@ mod tests {
         assert!(handler.handle_internal_request(&headers[1].id()).unwrap());
         assert!(!handler.handle_internal_request(&headers[1].id()).unwrap());
     }
+
+    //TODO(A0-2984): remove this after legacy sync is excised
```
Comment (kostekIV) [inline finality-aleph/src/sync/handler/mod.rs:1894]: we dont have to remove

File: finality-aleph/src/sync/handler/mod.rs [hunk 5]
```diff
@@ -82,6 +83,90 @@ pub trait HandlerTypes {
...
+                    break;
+                }
+                // we suppress all errors except `TooNew` since we are likely trying to mark things that are already marked and they would be throwing a lot of stuff
+                if let Err(ForestError::TooNew) = forest.update_body(&header) {
+                    *last_sync = top_finalized.number();
+                    return Ok(());
+                }
+                for child in chain_status.children(header.id())? {
+                    to_import.push_back(child);
+                }
```
Comment (kostekIV) [inline finality-aleph/src/sync/handler/mod.rs:162]: nit: use `to_import.extend`

File: finality-aleph/src/sync/handler/mod.rs (hunk 3 above)
Comment (timorl) [reply finality-aleph/src/sync/handler/mod.rs:173]: On closer examination probably no errors should be suppressed – this was something that I thought would be necessary in one of the previous approaches, and I didn't notice it no longer was in this one.

Changed, although unfortunately this means this should be properly tested again.

Had to introduce a new error type tho. ;-;

File: finality-aleph/src/sync/handler/mod.rs [hunk 6]
```diff
@@ -1790,4 +1890,35 @@ mod tests {
...
+            let justification = MockJustification::for_header(header.clone());
+            handler
+                .handle_justification_from_user(justification)
+                .expect("should work");
+            match notifier.next().await {
+                Ok(BlockFinalized(finalized_header)) => assert_eq!(
+                    header, &finalized_header,
+                    "should finalize the current header"
+                ),
+                _ => panic!("shoould notify about finalized block"),
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/sync/handler/mod.rs]: Fixed, both here and in the code I copied it from. :P

File: finality-aleph/src/sync/handler/mod.rs [hunk 7]
```diff
@@ -82,6 +83,90 @@ pub trait HandlerTypes {
...
+            AllGood => {
+                *self = MissedImports {
+                    highest_missed: missed,
+                    last_sync: chain_status.top_finalized()?.header().id().number(),
+                }
+            }
+            MissedImports { highest_missed, .. } => {
+                if *highest_missed < missed {
+                    *highest_missed = missed;
+                }
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/sync/handler/mod.rs]: Hah, done.

File: finality-aleph/src/sync/handler/mod.rs [hunk 8]
```diff
@@ -82,6 +83,90 @@ pub trait HandlerTypes {
     type Error;
 }
 
+// This is only required because we don't control block imports and thus we can get notifications about blocks being imported that don't fit in the forest. This struct lets us work around this by manually syncing the forest after such an event.
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/sync/handler/mod.rs]: Done, I recently switched to spacemax and apparently it doesn't do the splitting automatically the way I had vim configured. :<

File: finality-aleph/src/sync/handler/mod.rs (hunk 4 above)
Comment (timorl) [reply finality-aleph/src/sync/handler/mod.rs:1935]: We do actually – after removing the other parts the component will no longer satisfy the requirements of this test, but it also won't have to, as the only imports will be happening from the forest, thus never producing overzealous imports.

File: finality-aleph/src/sync/handler/mod.rs (hunk 5 above)
Comment (timorl) [reply finality-aleph/src/sync/handler/mod.rs]: Yup, done.

**Review by ggawryal — APPROVED** [review]

**Pull request merged**